### PR TITLE
ENH: infer Timestamp unit in non-iso paths

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/scalar/timestamp/test_constructors.py [hunk 1]
```diff
@@ -34,6 +34,14 @@ def test_construct_from_string_invalid_raises(self):
         with pytest.raises(ValueError, match="gives an invalid tzoffset"):
             Timestamp("200622-12-31")
 
+    def test_constructor_str_infer_reso(self):
+        # non-iso8601 path
+        ts = Timestamp("2016-01-01 1:30:01 PM")
+        assert ts.unit == "s"
+
+        ts = Timestamp("2016 June 3 15:25:01.345")
+        assert ts.unit == "ms"
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_constructors.py:43]: I think we need 3 tests here, one for each path:
- `dt = _parse_delimited_date(date_string, dayfirst, out_bestunit)`
- `_parse_dateabbr_string(`
- `dt = dateutil_parse(date_string, default=_DEFAULT_DATETIME,`

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/timestamp/test_constructors.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/timestamp/test_constructors.py]: good idea, updated

File: pandas/tests/scalar/timestamp/test_constructors.py [hunk 2]
```diff
@@ -34,6 +34,24 @@ def test_construct_from_string_invalid_raises(self):
         with pytest.raises(ValueError, match="gives an invalid tzoffset"):
             Timestamp("200622-12-31")
 
+    def test_constructor_str_infer_reso(self):
+        # non-iso8601 path
+
+        # _parse_delimited_date path
+        ts = Timestamp("2023/01/30")
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/timestamp/test_constructors.py:41]: this hits the iso path, you'll need
```
ts = Timestamp("01/30/2023")
```

also, would prefer to parametrise these - you can put `_parse_delimited_date path` (and the others) as `id` for the different cases

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/tests/scalar/timestamp/test_constructors.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/timestamp/test_constructors.py]: updated to use the suggested string.  didnt parametrize - on the margin id like to slow the growth of the pytest overhead

**Review by MarcoGorelli — APPROVED** [review]
Looks good, thanks @jbrockmendel

File: pandas/tests/scalar/timestamp/test_constructors.py (hunk 2 above)
Comment (MarcoGorelli) [reply pandas/tests/scalar/timestamp/test_constructors.py]: just out of interest, why?

Comment (jbrockmendel) [reply pandas/tests/scalar/timestamp/test_constructors.py]: pytest collects something like 200k tests and last time i checked that added up to more than 1GB before we even start executing any tests.  Similarly autouse fixtures get called for each test which means that much more overhead at runtime (i haven't measured this so assume it is a small-but-adds-up amount).

Doing a little bit more inside the tests and a little bit less outside can trim that footprint and overhead.  Adding a couple cases here isn't a big deal, but on the margin i try to get in the habit of leaning against parametrization when it isn't a compelling use case.

**Pull request merged**